src/openrct2/peep/Guest.cpp

```diff
@@ -4909,17 +4915,35 @@ void Guest::UpdateRideMazePathfinding()
         // Mark last edge as visited if previously in a dead-end or the previous node completly visited
         if (currentEntry.matchCoords(previousEntry) || previousEntry.isCompletlyVisited())
         {
-            currentEntry.markVisited(mazeLastEdge, openEdgesCount);
-            // Mark also previous chosen edge in case of dead-end-loop
-            if (currentEntry.matchCoords(previousEntry))
-                currentEntry.markVisited(maze_last_edge.atLastIntersection(), openEdgesCount);
+            // Offer amnesia to completly lost guest
+            if (currentEntry.isCompletlyVisited())
+            {
+                maze_pathfind_history.initialize();
+            }
+            else
+            {
+                currentEntry.markVisited(mazeLastEdge, openEdgesCount);
+                // Mark also previous chosen edge in case of dead-end-loop
+                if (currentEntry.matchCoords(previousEntry))
+                    currentEntry.markVisited(maze_last_edge.atLastIntersection(), openEdgesCount);
+            }
         }
 
-        // TODO: Reimplement 'wrong way' probability using entry index
-        openEdgesFlags &= ~currentEntry.getVisited();
+        // TODO: Tune these values
+        static const std::array<std::array<uint16_t, 4>, 6> rememberThreshold = { {
+            // visited,           visited+peeked,     origin,             origin+peeked
+            { maze_set_odd(0.04), maze_set_odd(0.00), maze_set_odd(0.02), maze_set_odd(0.00) }, // Last intersection
+            { maze_set_odd(0.09), maze_set_odd(0.03), maze_set_odd(0.03), maze_set_odd(0.01) }, // -1
+            { maze_set_odd(0.27), maze_set_odd(0.07), maze_set_odd(0.18), maze_set_odd(0.10) }, // -2
+            { maze_set_odd(0.39), maze_set_odd(0.15), maze_set_odd(0.25), maze_set_odd(0.17) }, // -3
+            { maze_set_odd(0.50), maze_set_odd(0.31), maze_set_odd(0.45), maze_set_odd(0.32) }, // -4
```

Comment:
Why is it that visiting is less effective than origin, but visited+peeked is more effective than origin+peeked a couple intersections back?  I would expect the relation between the two to be related, ie either you tend to forget your origin faster than your visited or peeking helps them both equally.

Reply:
You right, the odds relations are her inconstant and I fixed that in 65573e6.
Regarding the fact that guests tends to forget less easily the origin than other visited directions, this is to encourage them to explore completely an area before turning back and so avoiding multiples travels between 2 area of the maze.